### DISCUSSION TITLE
How to call JS function from C (or get exported value from `JS_Eval`)?

### DISCUSSION THREAD

**Discussion by guyutongxue** [quickjs-ng/quickjs#643] [Q&A] [answered]
Hello! I'm new to QuickJS and I want to use it to embed my JS project into a C-API. But I'm encountering a problem that I don't know how to call JS functions in a `JS_Eval`ed scripts. For example:

```cpp
auto SOURCE = R"(
export function hello() {
  return 'Hello';
}
)" "\0"sv;

int main() {
  auto rt = JS_NewRuntime();
  auto ctx = JS_NewContext(rt);
  auto value = JS_Eval(ctx, SOURCE.data(), SOURCE.size() - 1, "main.js",  JS_EVAL_TYPE_MODULE);
  //   ^- Seems to be Promise<undefined> ?
  JSModuleDef* def = /* ??? */;
  auto ns = JS_GetModuleNamespace(ctx, def);
  auto fn = JS_GetPropertyStr(ctx, ns, "hello");
  auto result = JS_Call(ctx, fn, JS_UNDEFINED, 0, nullptr);
  auto str = JS_ToCString(ctx, result); // Expected "Hello"
  // [...]
  JS_FreeContext(ctx);
  JS_FreeRuntime(rt);
}
```

It's that possible to get the `export`ed value from a `JS_Eval`ed module? (The `???` part on above code) Or there is some other way to call JS functions from C code?

**Comment by bnoordhuis** [comment]
I can't guarantee it'll work forever but the return value from JS_Eval is the module, and that's just the wrapped JSModuleDef pointer:
```c
JSValue v = JS_Eval(ctx, buf, len, filename, JS_EVAL_TYPE_MODULE);
if (JS_VALUE_GET_TAG(v) == JS_TAG_MODULE) {
    JSModuleDef *m = JS_VALUE_GET_PTR(v);
    // ...
}
```

**Reply by guyutongxue** [reply]
🤔 Thanks, but the `JS_VALUE_GET_TAG(v)` returns `JS_TAG_OBJECT` not `JS_TAG_MODULE` ...

**Reply by bnoordhuis** [reply]
Sorry, forgot to mention you also need to pass `JS_EVAL_FLAG_COMPILE_ONLY` to JS_Eval.

But, now that I think of it, you'll end up with a JSModuleDef that isn't linked or fully evaluated yet, so how useful that is remains to be seen.

**Reply by saghul** [reply]
You'd need to `JS_EvalFunction` the module and use `JS_GetModuleNamespace` to get the namespace object.

**Comment by guyutongxue** [comment, marked as answer]
Thank you @bnoordhuis and @saghul ! Here is a working example (sorry for using C++):

```cpp
int main() {
  auto rt = JS_NewRuntime();
  auto ctx = JS_NewContext(rt);

  auto bytecode = JS_Eval(ctx, SOURCE.data(), SOURCE.size() - 1, "main.js",
                          JS_EVAL_TYPE_MODULE | JS_EVAL_FLAG_COMPILE_ONLY);
  JS_ResolveModule(ctx, bytecode);
  JS_EvalFunction(ctx, bytecode);
  auto def = static_cast<JSModuleDef*>(JS_VALUE_GET_PTR(bytecode));
  auto ns = JS_GetModuleNamespace(ctx, def);

  auto fn = JS_GetPropertyStr(ctx, ns, "hello");
  auto result = JS_Call(ctx, fn, JS_UNDEFINED, 0, nullptr);
  auto str = JS_ToCString(ctx, result);
  std::cout << str << std::endl;  // Expected "Hello"

  JS_FreeContext(ctx);
  JS_FreeRuntime(rt);
}
```

Use `JS_EVAL_FLAG_COMPILE_ONLY` to get a `JSModuleDef` and use `JS_ResolveModule` (I'm not sure if this is necessary but anyway it works) and `JS_EvalFunction` to fully load this module. Then `JS_GetModuleNamespace` works like a charm.

**Reply by saghul** [reply]
great! Yeah I _think_ you can remove `JS_ResolveModule` it's likely called internally anyway, but I'm not 100% sure.

**Comment by dbalabka** [comment]
@saghul, how do we await the function result if the function is `async`? Thanks in advance!
```
export async function hello() {
  return 'Hello';
}
```

**Reply by bnoordhuis** [reply]
The return value when you call an async function is a promise. Call `JS_ExecutePendingJob` until `JS_PromiseState` no longer returns `JS_PROMISE_PENDING`.
